storage/innobase/xtrabackup/src/xbcrypt.c

```diff
@@ -400,8 +395,27 @@ mode_decrypt(File filein, File fileout)
 			decryptbuf = chunkbuf;
 		}
 
+		if (hash_appended) {
+			uchar hash[XB_CRYPT_HASH_LEN];
+
+			originalsize -= XB_CRYPT_HASH_LEN;
+
+			assert(gcry_md_get_algo_dlen(XB_CRYPT_HASH) ==
+			       XB_CRYPT_HASH_LEN);
+			gcry_md_hash_buffer(XB_CRYPT_HASH, hash, decryptbuf,
+					    originalsize);
+			if (memcmp(hash, (char *) decryptbuf + originalsize,
+				   XB_CRYPT_HASH_LEN) != 0) {
+				msg("%s:%s invalid plaintext hash. "
+				    "Wrong encrytion key specified?\n",
+				    my_progname, __FUNCTION__);
+				result = XB_CRYPT_READ_ERROR;
+				goto err;
+			}
+		}
+
 		/* Write it out */
-		if (my_write(fileout, decryptbuf, originalsize,
+		if (my_write(fileout, (const uchar *) decryptbuf, originalsize,
```

Comment:
actually, this buffer is used in two places. We pass it to libgrypt and we pass it to `my_write`. format of this buffer is binary, no encodings involved. We output the buffer "as is" after it was processed by libgrypt. Issue is, `libgcrypt` declares buffers as `void *` while `my_write` takes `uchar *`. So, we need a type cast to make compiler happy.